File: c/fannkuch-redux/fannkuchredux.c
#include <stdio.h>
#include <stdlib.h>

int fannkuch(int n) {
    int *perm  = malloc(n * sizeof(int));
    int *perm1 = malloc(n * sizeof(int));
    int *count = malloc(n * sizeof(int));
    int maxFlips = 0, permCount = 0, checksum = 0;

    // Initialize perm1 with 1 to n
    for (int i = 0; i < n; i++) {
        perm1[i] = i;
    }

    int r = n;

    while (1) {
        while (r != 1) {
            count[r-1] = r;
            r--;
        }

        // Copy perm1 to perm
        for (int i = 0; i < n; i++) {
            perm[i] = perm1[i];
        }

        int flipsCount = 0;
        int k;

        while ((k = perm[0]) != 0) {
            // k is the first number in the array, so we flip perm[0]...perm[k]
            int k2 = (k + 1) >> 1;
            for (int i = 0; i < k2; i++) {
                int temp = perm[i];
                perm[i] = perm[k - i];
                perm[k - i] = temp;
            }
            flipsCount++;
        }

        maxFlips = flipsCount > maxFlips ? flipsCount : maxFlips;
        permCount++;

        // Generate next permutation
        while (1) {
            if (r == n) {
                free(perm);
                free(perm1);
                free(count);
                return maxFlips;
            }

            int perm0 = perm1[0];
            int i = 0;
            while (i < r) {
                int j = i + 1;
                perm1[i] = perm1[j];
                i = j;
            }
            perm1[r] = perm0;

            count[r] = count[r] - 1;
            if (count[r] > 0) break;
            r++;
        }
    }
}

int main() {
    int n = 12;
    printf("Maximum flips for n=%d is %d\n", n, fannkuch(n));
    return 0;
}

File: fortran/fannkuch-redux/fannkuchredux.f90
program fannkuch

  implicit none

  integer :: n,checksum,maxFlipsCount
  character(len=2) :: arg
  character(len=10) :: out
  call get_command_argument(1,arg);read(arg,'(I2)')n
  call fannkuchredux(n,checksum,maxFlipsCount)
  write(out,'(I10)') checksum;
  write(*,'(A)') trim(adjustl(out))
  write(out,'(I10)') maxFlipsCount;
  write(*,'(A)') 'Pfannkuchen('//trim(adjustl(arg))//') = '//trim(adjustl(out))

contains

  subroutine fannkuchredux(n,checksum,maxFlipsCount)
    implicit none
    integer, intent(in) :: n
    integer, intent(out) :: maxFlipsCount,checksum
    integer :: perm(0:n-1),perm1(0:n-1),icount(0:n-1);
    integer :: i,r,temp,k,perm0,permCount = 0,flipsCount = 0;
    checksum = 0;  maxFlipsCount = 0;
    do i = 0,n-1
       perm1(i)= i;
    enddo
    r=n
    do while (.true.)
       do while (r /= 1)
          icount(r-1)=r; r=r-1;
       end do
       perm = perm1; flipsCount = 0;
       do while (perm(0)/=0) 
          k = perm(0)
          do  i=0,ishft(k+1,-1)-1
             temp = perm(i); perm(i) = perm(k-i); perm(k-i) = temp;
          enddo
          flipsCount = flipsCount + 1;
       end do
       maxFlipsCount = max(maxFlipsCount,flipsCount);
       if (modulo(permCount,2) == 0)then
          checksum = checksum + flipsCount
       else
          checksum = checksum - flipsCount
       end if
       do while (.true.)
          if (r==n) return
          perm0 = perm1(0); i = 0;
          do while (i < r) 
             perm1(i) = perm1(i + 1); i = i + 1;
          end do
          perm1(r) = perm0;
          icount(r) = icount(r) - 1;
          if (icount(r) > 0) exit;
          r = r + 1;
       end do
       permCount = permCount + 1;
    end do
  end subroutine fannkuchredux
  
end program fannkuch